README
Business-PayPal-API version 0.62
================================

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   WPP_TEST=auth.txt make test
   make install

Please notice that this module requires you have several things before
you can test it:

  - a sandbox personal PayPal account
  - a sandbox business PayPal account
  - API credentials (either a certificate or signature)
  - auth.txt, which contains your API credentials

Acquiring these things is your job, not mine. Read PayPal's
and this module's documentation to learn more.

If you do not set the WPP_TEST environment variable, sandbox tests
will be skipped.

The format of the authentication tokens file defined by WPP_TEST may
be found in the Business::PayPal::API documentation under
"TESTING". Sample auth.txt files may be found in 'auth.sample.3token'
and 'auth.sample.cert' in this distribution.

USAGE

  use Business::PayPal::API qw( ExpressCheckout );

  my $pp = new Business::PayPal::API( Username => 'my_api.username.tld',
                                      Password => 'API_PASSWORD',
                                      CertFile => '/path/to/cert.pem',
                                      KeyFile  => '/path/to/key.pem', );

  ...
  my %details = $pp->GetExpressCheckoutDetails( $token );

DEPENDENCIES

This module requires these other modules and libraries:

  SOAP::Lite 0.67 or later.
  Crypt::SSLeay required for certificate auth

COPYRIGHT AND LICENCE

This package except those modules listed below are copyright (C) 2006
by Scott Wiersdorf

AuthorizationRequest, CaptureRequest, DirectPayments,
ReauthorizationRequest, and VoidRequest modules are copyright (C) 2006
by Daniel Hembree

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.6 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Business::PayPal::API.

- Merge in changes from Ron Phipps.
  - Reduce SOAP version requirement.
  - Protection eligibility support.
- Merge in changes from Tomas Doran.
  - CPAN packaging enhancements.
- Extract out more payer details from XML. (PayerName, NameSuffix, PayerCountry).
- Fix warning in Perl 5.14.x (hopefully). https://rt.cpan.org/Public/Bug/Display.html?id=70424
- Fix https://rt.cpan.org/Public/Bug/Display.html?id=67386

Release 0.69 - As currently release to CPAN
-------------------------------------------

(Changelog corrected).

Release 0.68 - First hembreed release
-------------------------------------
I've moved the module's scm to http://github.com/hembreed/Business-PayPal-API so others may share my pain. CPAN will be updated from there.

- API.pm
   1. Changed eq 'Success' to =~ /Success/ to accomodate variations introduced
      by PayPal.
   2. Changed test for duplicates to test for defined variable first. This
      was causing errors in 5.10+ perl versions.
-t/*t Changed test for Success in all test modules as above
-t/RecurringPayments Corrected errors in type definitions and testing. Needs more work.
-TransactionSearch Corrected test instructions and code to allow transaction searches using the transactionID. This works for some types of transactions but not others as per PayPal. It is   the only way to get MassPayment information.

Release 0.62 - last scottw release
----------------------------
revision 1.24
date: 2009/07/28 18:00:58;  author: scott;  state: Exp;  lines: +44 -4
- getFields() returns an array ref if multiple values are returned
  from PayPal, otherwise it behaves as it always did, returning a
  non-ref scalar.
- BillingAgreement information returned in ExpressCheckout
- Fix GetTransactionDetails 'PII_SalesTax' parameter
- Fix GetTransactionDetails to return multiple PaymentItems
- Fix MassPay to use the correct UniqueID parameter
- Add DoReferenceTransaction to RecurringPayments API
- Lots of credit given to great folks with patches in API docs

Thanks everyone! I'm handing this module off to Danny Hembree for
maintenance now.

Release 0.61
----------------------------
revision 1.23
date: 2008/05/05 15:10:40;  author: scott;  state: Exp;  lines: +37 -5
- timeout setting available
- additional troubleshooting documentation (auth error handling, URL munging)
----------------------------
revision 1.4 [DirectPayments.pm]
date: 2008/05/05 15:11:14;  author: scott;  state: Exp;  lines: +34 -2
- add shipto parameters
----------------------------
revision 1.12 [ExpressCheckout.pm]
date: 2008/05/05 15:11:51;  author: scott;  state: Exp;  lines: +11 -3
- MaxAmount parameter take currencyID attribute (Sandbox)
----------------------------
revision 1.1 [RecurringPayments.pm]
date: 2008/05/05 15:15:04;  author: scott;  state: Exp;
- add base for RecurringPayments (not working yet)

Release 0.51
----------------------------
revision 1.22
date: 2007/09/27 20:32:31;  author: scott;  state: Exp;  lines: +18 -6
- [API.pm] add three-token signature testing URI (Oliver Ready)
- [DirectPayments.pm] add ShippingTotal field (patch: Michael Hendricks)
- [VoidRequest.pm] documentation fixes for method call (Oliver Ready)
- [ReauthorizationRequest.pm] documentation fixes for method call (Oliver Ready)

Release 0.51
----------------------------
revision 1.10
date: 2007/08/29 20:56:42;  author: scott;  state: Exp;  lines: +4 -2
- fix PayerStatus location in DOM, add AddressStatus element (patch: Michael Hendricks)

Release 0.50
----------------------------
revision 1.20
date: 2007/05/21 21:59:33;  author: scott;  state: Exp;  lines: +5 -5
- minor doc changes
- new Mass Pay API module included
----------------------------

Release 0.41
----------------------------
revision 1.19
date: 2007/05/15 20:46:56;  author: scott;  state: Exp;  lines: +34 -12
- some minor documentation additions

Release 0.40
----------------------------
revision 1.18
date: 2006/10/06 17:53:44;  author: scott;  state: Exp;  lines: +2 -2
- version bump
----------------------------
revision 1.17
date: 2006/10/06 17:49:50;  author: scott;  state: Exp;  lines: +42 -38
- using PayPal API version 2.0
- note about SSL requirements of LWP
- minor documentation cleanup
- DirectPayments tests added
- New modules (with tests!) from Danny Hembree:
  AuthorizationRequest.pm
  CaptureRequest.pm
  ReauthorizationRequest.pm
  VoidRequest.pm
----------------------------
revision 1.16
date: 2006/07/06 15:25:21;  author: scott;  state: Exp;  lines: +9 -2
- add acknowledgement and pointers to example code from Andy Spiegl and others

Release 0.33
----------------------------
revision 1.15
date: 2006/07/05 18:05:39;  author: scott;  state: Exp;  lines: +2 -2
- fix documentation regarding SetExpressCheckout (returns a hash, not
  a scalar). If you were following the code in the SYNOPSIS for
  ExpressCheckout and getting a token with '3/8' or '4/8', you should
  change the $token to %response, since SetExpressCheckout() returns a
  hash (big thanks to Andy Spiegl for finding this).

Release 0.32
----------------------------
revision 1.14
date: 2006/07/03 15:46:24;  author: scott;  state: Exp;  lines: +16 -1
- acknowledgements section
----------------------------
revision 1.13
date: 2006/07/03 15:40:16;  author: scott;  state: Exp;  lines: +11 -2
- fix for 3-token auth

Release 0.31
----------------------------
revision 1.12
date: 2006/06/29 02:36:24;  author: scott;  state: Exp;  lines: +2 -2
- fix StateOrProvince typo [rt.cpan.org #20176]

Release 0.30
----------------------------
revision 1.11
date: 2006/04/18 16:23:18;  author: scott;  state: Exp;  lines: +2 -2
- added DirectPayments API by Daniel Hembree. Thanks Daniel!

Release 0.23
----------------------------
revision 1.10
date: 2006/04/04 19:29:08;  author: scott;  state: Exp;  lines: +12 -5
- typo fix
- wrap soap call in eval for safety (if ssl neg. fails w/ paypal, it croaks)
- version bump

Release 0.22

lib/Business/PayPal/API/TransactionSearch.pm
package Business::PayPal::API::TransactionSearch;

use 5.008001;
use strict;
use warnings;

use SOAP::Lite 0.67;
use Business::PayPal::API ();

our @ISA = qw(Business::PayPal::API);
our $VERSION = '0.11';
our $CVS_VERSION = '$Id: TransactionSearch.pm,v 1.3 2006/03/24 17:11:37 scott Exp $';
our @EXPORT_OK = qw( TransactionSearch );

sub TransactionSearch {
    my $self = shift;
    my %args = @_;

    my %types = ( StartDate        => 'xs:dateTime',
                  EndDate          => 'xs:dateTime',
                  Payer            => 'ebl:EmailAddressType',
                  Receiver         => 'ebl:EmailAddressType',
                  ReceiptID        => 'xs:string',
                  TransactionID    => 'xs:string',
                  InvoiceID        => 'xs:string',
                  PayerName        => 'xs:string',
                  AuctionItemNumer => 'xs:string',
                  TransactionClass => '',
                  Amount           => 'ebl:BasicAmountType',
                  CurrencyCode     => 'xs:token',
                  Status           => '',
                );

    my @trans = 
      (
       $self->version_req,
       SOAP::Data->name( StartDate => $args{StartDate} )->type( delete $types{StartDate} )
      );

    for my $type ( keys %types ) {
        next unless $args{$type};
        push @trans, SOAP::Data->name( $type => $args{$type} )->type($types{$type});
    }

    my $request = SOAP::Data->name
      ( TransactionSearchRequest => \SOAP::Data->value( @trans ) )
	->type("ns:TransactionSearchRequestType");

    my $som = $self->doCall( TransactionSearchReq => $request )
      or return;

    my $path = '/Envelope/Body/TransactionSearchResponse';

    my %response = ();
    unless( $self->getBasic($som, $path, \%response) ) {
        $self->getErrors($som, $path, \%response);
        return %response;
    }

    return $self->getFieldsList( $som, $path . '/PaymentTransactions',
                                 { Timestamp        => 'Timestamp',
                                   Timezone         => 'Timezone',
                                   Type             => 'Type',
                                   Payer            => 'Payer',
                                   PayerDisplayName => 'PayerDisplayName',
                                   TransactionID    => 'TransactionID',
                                   Status           => 'Status',
                                   GrossAmount      => 'GrossAmount',
                                   FeeAmount        => 'FeeAmount',
                                   NetAmount        => 'NetAmount',
                                 } );
}

1;
__END__

=head1 NAME

Business::PayPal::API::TransactionSearch - PayPal TransactionSearch API

=head1 SYNOPSIS

  use Business::PayPal::API::TransactionSearch;

  ## see Business::PayPal::API documentation for parameters
  my $pp = new Business::PayPal::API::TransactionSearch ( ... );

  my %response = $pp->TransactionSearch( StartDate => '1998-01-01T00:00:00Z',
                                         TransactionID => $transid, );

=head1 DESCRIPTION

B<Business::PayPal::API::TransactionSearch> implements PayPal's
B<TransactionSearch> API using SOAP::Lite to make direct API calls to
PayPal's SOAP API server. It also implements support for testing via
PayPal's I<sandbox>. Please see L<Business::PayPal::API> for details
on using the PayPal sandbox.

=head2 TransactionSearch

Implements PayPal's B<TransactionSearch> API call. Supported
parameters include:

  StartDate (required)
  EndDate
  Payer
  Receiver
  TransactionID
  PayerName
  AuctionItemNumber
  InvoiceID
  TransactionClass
  Amount
  CurrencyCode
  Status

as described in the PayPal "Web Services API Reference" document. The
syntax for StartDate is:

  YYYY-MM-DDTHH:MM:SSZ

'T' and 'Z' are literal characters 'T' and 'Z' respectively, e.g.:

  2005-12-22T08:51:28Z

Returns a list reference containing up to 100 matching records (as per
the PayPal Web Services API). Each record is a hash reference with the
following fields:

  Timestamp
  Timezone
  Type
  Payer
  PayerDisplayName
  TransactionID
  Status
  GrossAmount
  FeeAmount
  NetAmount

Example:

  my $records = $pp->TransactionSearch( StartDate => '2006-03-21T22:29:55Z',
                                        InvoiceID => '599294993', );
  
  for my $rec ( @$records ) {
      print "Record:\n";
      print "TransactionID: " . $rec->{TransactionID} . "\n";
      print "Payer Email: " . $rec->{Payer} . "\n";
      print "Amount: " . $rec->{GrossAmount} . "\n\n";
  }

=head2 ERROR HANDLING

See the B<ERROR HANDLING> section of B<Business::PayPal::API> for
information on handling errors.

=head2 EXPORT

None by default.

=head1 SEE ALSO

L<https://developer.paypal.com/en_US/pdf/PP_APIReference.pdf>

=head1 AUTHOR

Scot Wiersdorf E<lt>scott@perlcode.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2006 by Scott Wiersdorf

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.5 or,
at your option, any later version of Perl 5 you may have available.


=cut
